### ISSUE TITLE
Tree flipping algorithm

### ISSUE THREAD

**Issue by bkamins** [bkamins/SilverDecisions#42]
In the future it would be nice to have tree-flipping functionality implemented.

[Here](https://www.overleaf.com/read/yjghksqjfsms) we maintain with @michaljakubczyk the specification of the desired functionality.

**Comment by bkamins** [comment]
When the tree is flipped we accept that we have to normalize probabilities to sum to 1.
This means that "double flip" is not guaranteed to revert us to original probabilities (which again, in particular, means that in some corner cases optimal path might change - but this should be very rare).

In general after a flip probabilities should be normalized to sum to 1. After a thought - WE DO NOT DO ROUNDING BUT ACCEPT IT (if it is convenient to do in implementation), only normalization should be guaranteed.

@mwasiluk let @bkamins know how you implemented it so that I can document it properly.

@annawier remember to put the latex specification to documentation.

**Comment by bkamins** [comment]
If you flip this tree:
<img width="418" alt="start" src="https://cloud.githubusercontent.com/assets/6187170/21026331/6aaac23c-bd8c-11e6-9f1c-8aae987c3094.png">

1) after "Flip tree" action context menu stays and nothing happens
2) the tree is flipped in the background, and repaints if you click e.g. decision node
3) however, the repainted tree is rendered incorrectly (the child chance nodes are painted too small)
4) probabilities in lower (`b`) branch to not sum to 1 (but to 0)
5) but the tree does not report is as an error with `!!`
6) however, a reverse-flip is impossible

@michaljakubczyk in conclusion: we need heavy testing of this. It is your request so ...
How to test:
1) go to SilverDecisions <code> section
2) select Branch dev (by default it is master)
3) click "clone or download" and download as ZIP
4) you have whole repository in ZIP with "node injection" and "tree-flipping"

**Comment by bkamins** [comment]
I will update documentation to reflect equal distribution of probabilities in corner case.

**Comment by bkamins** [comment]
@michaljakubczyk please check the added description and the resulting behavior in SilverDecisions - maybe you prefer something else (and again - detailed testing is required here).

**Comment by bkamins** [comment]
@mwasiluk Curious behavior of tree flipping (not a bug but I want to understand why this happens):

Initial tree structure (edges are ordered `b/d/c`):
<img width="410" alt="example1" src="https://cloud.githubusercontent.com/assets/6187170/21030109/203235d0-bd9d-11e6-9df3-f16c2729f070.png">


1) I flip it
2) I move edge `c` to top (over 'b') in both branches
3) I flip it and the order `b/d/c` is preserved (and I would expect order `c/b/d` and I would get it if the tree after step 2) were drawn manually)

A more general question (for documentation): how do you decide how to order edges after the flip if in the subtrees the edges are not arranged in the same way? E.g. in the picture above in first branch we could have `x/y` order and in the second branch `y/x` order.

**Comment by mwasiluk** [comment]
@bkamins it was a bug. Edges were not always sorted (by y location) in the model after parent node was dragged. They were sorted only before displaying in the left sidebar. Fixed.

Now edges in the subtrees must be arranged in the same way to allow flipping. 
Maybe it could be changed that e.g. order is decided by the first subtree?

**Comment by bkamins** [comment]
After 1c136fc it is OK and I understand we have the following functionality
- tree flip correctly recognizes order of children of child chance nodes
- tree flip is allowed only if in all child nodes the order is the same (i.e. we disallow permutations of labels)

@michaljakubczyk when we go to production please document this functionality on Wiki (@annawier does not know it well enough).
